src/ray_tracer.cpp

```diff
@@ -90,36 +90,103 @@ RayTraceIterRange rayTrace (const nm::MapMetaData& info, const gm::Point& p1, co
   return rayTrace(c1, c2);
 }
 
-
-gm::Point rayEndPoint (const gm::Point& p0, const double theta, const double d)
+// Singleton class to generate points at max range of each ray
+// of a scan from a specified sensor pose
+// Caches the points and and as long as the scan properties are
+// the same, only a transformation to the new sensor pose is applied
+class ScanEndPoints
 {
-  gm::Point p;
-  p.x = p0.x + cos(theta)*d;
-  p.y = p0.y + sin(theta)*d;
-  return p;
-}
+private:
+  ScanEndPoints()
+    : angle_min_(0.0)
+    , angle_max_(0.0)
+    , range_max_(0.0)
+    , angle_increment_(0.0)
+  {}
+  ScanEndPoints(const ScanEndPoints&);
+  void operator=(const ScanEndPoints&);
+
+public:
+  typedef std::vector<gm::Point>::const_iterator PointIter;
+  typedef std::pair<PointIter, PointIter> PointRange;
+
+  static ScanEndPoints& getInstance()
+  {
+    static ScanEndPoints instance;
+    return instance;
+  }
+
+  PointRange operator()(const sm::LaserScan& scanner_info, const gm::Pose& sensor_pose)
+  {
+    validate(scanner_info);
+    transform(sensor_pose);
+    return std::make_pair(points_.begin(), points_.end());
+  }
+
+private:
+  void validate(const sm::LaserScan& info)
+  {
+    if (angle_min_ == info.angle_min && angle_max_ == info.angle_max &&
```

Comment:
Another option is to compare only the `info.header.frame_id`, but I see it's orthogonal to this, as it affects the sensor pose. But in general, just checking the `frame_id` you can assume that even the `points_` don't change. You can do that on another PR instead of this one. IMHO it'd make this even simpler, faster and it's a valid assumption since the LIDARs usually have a static frame_id. You can also try to support static and moving LIDARs. Of course, on another PR.